src/bin/cargo.rs

```diff
@@ -69,7 +68,31 @@ See 'cargo help <command>' for more information on a specific command.
 
 fn main() {
     env_logger::init().unwrap();
-    execute_main_without_stdin(execute, true, USAGE)
+
+    let config = match Config::default() {
+        Ok(cfg) => cfg,
+        Err(e) => {
+            let mut shell = cargo::shell(Verbosity::Verbose, ColorConfig::Auto);
+            cargo::handle_cli_error(e.into(), &mut shell);
+            return;
+        }
+    };
+
+    let result = (|| {
+        let args: Vec<_> = try!(env::args_os().map(|s| {
+            s.into_string().map_err(|s| {
+                human(format!("invalid unicode in argument: {:?}", s))
+            })
+        }).collect());
+        let rest = &args;
+        cargo::call_main_without_stdin(execute, &config, USAGE, rest, true)
+    })();
+
+    match result {
+        Err(e) => cargo::handle_cli_error(e, &mut *config.shell()),
+        Ok(None) => {},
+        Ok(Some(())) => unreachable!(),
```

Comment:
This case is here because we have a great infrastrucure for commands to return `Encodable` instances to be printing in the JSON format. 

In reality, only three commands use this feature (deprecated `read_manifest`, `verify_project` which prints `Map<String, String>` of errors and `metadata`). Everyone else is forced to return the funny `Option<()>` type. I want to strip this infrastructure in the next PR and instead make the three commands to print JSON directly.